### PR TITLE
Experimental theme package

### PR DESCRIPTION
**Pull request by JasonVMo**
### Platforms Impacted
- [ ] iOS
- [ ] macOS
- [x] win32 (Office)
- [ ] windows
- [ ] android

### Description of changes

This is both a refactoring change and a proposal of an alternative way of handling theming from a customer perspective.

#### `@fluentui-react-native/theme` module

This creates a new class called `ThemeReference` which is used to wrap theme objects. This class provides:

- the ability to layer a base theme, or another `ThemeReference`, together with a set of recipes to create a new theme. These recipes are either partial themes, or functions that produce partial themes
- an update mechanism, to listen for changes
- the ability to parent themes to other themes and have updates propogate as expected

This is designed to replace the code in theming-ramp, theme-registry, and theming-react-native (though that required factoring out the native module).

The typing on this is loose, the idea being that people could put their own variants of the theme and pass it through our context. Commenting for instance could extend our theme definition to add their own additional values, then look up information from their native module in a functional handler to merge the two approaches.

This also creates a new `ThemeProvider` which is a bit simpler as it only needs theme, not a registry.

- this shares the same ThemeContext as existing providers
- components will continue to work with either approach as they only reference the context for their lookups 

#### `@fluentui-react-native/win32-theme` module

This moves the code to produce an Office theme, including the code that works with the native module, into its own package. 

- This lets us take that code out of the bundle for platforms that don't need it.
- It opens the door for the native modules to vary slightly between platforms
- Customers that want their own native module can bypass any of our native module based solutions and just use our default theme

#### Usage

In practice the usage looks like:

```tsx
import { createOfficeTheme } from '@fluentui-react-native/win32-theme';
import { ThemeProvider, ThemeReference } from '@fluentui-react-native/theme';

// this will actually create a theme reference
const theme = createOfficeTheme('TaskPane');

// create a slight variant, based off the Office theme (this could be wrapped in a function as well)
const variantTheme = new ThemeReference(theme, parent => {
  // do some stuff
});

const render = () => {
  <ThemeProvider theme={theme}>
    //content
  </ThemeProvider>
}
```

### Verification

Did the following:

- added some UTs to the theme package
- ran through the web tester manually
- ran through the win32 tester side by side with an unchanged version

| Before                                       | After                                      |
|----------------------------------------------|--------------------------------------------|
| Screenshot or description before this change | Screenshot or description with this change |

### Pull request checklist

This PR has considered (when applicable):
- [x] Automated Tests
- [ ] Documentation and examples
- [ ] Keyboard Accessibility
- [ ] Voiceover
- [ ] Internationalization and Right-to-left Layouts

### CODE REVIEW

File: packages/experimental/theme/README.md
Comment (tkamireh) [inline packages/experimental/theme/README.md:3]: building and layering?

File: packages/experimental/theme/README.md
Comment (tkamireh) [inline packages/experimental/theme/README.md:56]: If I'm understanding this correctly, what `stagedComponent` give us is a way to compress from n (where n > 2) layered components in the render tree to 2. So if `BaseComponent` renders a `View` in it's continuation `WrapperComponent` looks like
```
<WrapperComponent>
      <View>
            ....
      </View>
</WrapperComponent>
```

So is this pulling out some of the ideas from the original `compose` and `composable` concepts into their own helpers?

File: packages/experimental/theme/src/mergeTheme.ts
Comment (tkamireh) [inline packages/experimental/theme/src/mergeTheme.ts:5]: is the casts to `unknown` necesary - and does immutableMerge not accept partials (or maybe TS isn't smart enough to realize it's a partial)?

File: packages/theming/win32-theme/src/createPartialOfficeTheme.ts
Comment (tkamireh) [inline packages/theming/win32-theme/src/createPartialOfficeTheme.ts:7]: I've seen `index` a few times, what does it get used for? And why do we default to -1?

File: packages/experimental/theme/README.md
Comment (JasonVMo) [reply packages/experimental/theme/README.md:3]: Doh, just copied this file (obviously) and need to update this.

File: packages/experimental/theme/README.md
Comment (JasonVMo) [reply packages/experimental/theme/README.md:56]: Yes, that's the way the experimental framework is built. The pieces can be used on their own, or together. Compose brings them all together, there is no more vanilla composable (which was confusing to folks).

That being said, this is the existing documentation file in use-slots, which accidentally got copied when I created the new package.

File: packages/experimental/theme/src/mergeTheme.ts
Comment (JasonVMo) [reply packages/experimental/theme/src/mergeTheme.ts:5]: TS isn't smart enough because a Partial<MyType> is just a mapped type where it adds ? to every key within. The theme partial type is built up from explicitly declared partial sub-types. You can build a deep partial but then it does weird things to objects like ColorRamps where you really should be specifying multiple elements.

File: packages/theming/win32-theme/src/createPartialOfficeTheme.ts
Comment (JasonVMo) [reply packages/theming/win32-theme/src/createPartialOfficeTheme.ts:7]: This was actually moved from theming-react-native, @FalseLobster might have better context for that.

File: packages/experimental/theme/README.md
Comment (tkamireh) [reply packages/experimental/theme/README.md:56]: Makes sense :)

File: packages/experimental/theme/src/mergeTheme.ts
Comment (tkamireh) [reply packages/experimental/theme/src/mergeTheme.ts:5]: Ah right the deep partials are what you want

**Review by tkamireh — APPROVED** [review]
Besides documentation changes, lgtm

File: packages/theming/win32-theme/src/createPartialOfficeTheme.ts
Comment (FalseLobster) [reply packages/theming/win32-theme/src/createPartialOfficeTheme.ts:7]: That was based on trying to align with a now out-of-date version of the Web's theming ideas.  I don't think it's relevant any more and should probably get removed since it's just complicating the theme dictionary.

File: packages/framework/theming-react-native/src/NativeModule/office/debugpalette.json
Comment (FalseLobster) [inline packages/framework/theming-react-native/src/NativeModule/office/debugpalette.json:1]: Did this get deleted and not just moved?

File: packages/theming/win32-theme/src/createPartialOfficeTheme.ts
Comment (JasonVMo) [reply packages/theming/win32-theme/src/createPartialOfficeTheme.ts:7]: I agree, though I want the design we move towards to be a bit more stable before we transition.

File: packages/framework/theming-react-native/src/NativeModule/office/debugpalette.json
Comment (JasonVMo) [reply packages/framework/theming-react-native/src/NativeModule/office/debugpalette.json:1]: Yes. Looking at it I don't think we need it since this was only being used in the case where the native module was not being used, but we already have a complete palette definition with the white palette so this seemed redundant.

Comment (FalseLobster) [reply packages/framework/theming-react-native/src/NativeModule/office/debugpalette.json:1]: I think you're overlooking a few things here, but I'm still trying to run through these changes to make sure I understand the repercussions, so give me a moment to collect my thoughts

Comment (FalseLobster) [reply packages/framework/theming-react-native/src/NativeModule/office/debugpalette.json:1]: I think my main concern is that the fallback module doesn't actually have a full palette whereas the debugpalette.json did.  This means that certain keys in the office palette object will be undefined when consumers turn on web debugging.  This will likely break existing consumers even if the Fluent theme translation only uses a subset of the native colors.

It's also worth noting that just because getPalette (a synchronous native module) is unsafe in web debugger, that doesn't mean the module isn't there.  The old approach just replaced the actual NM.getPalette() call with a web-debug safe version but left the rest of the NM intact, whereas the new one actually always uses the fallback Module.

The recipe based approach does look much nicer over all and simplifies things, though.

Comment (JasonVMo) [reply packages/framework/theming-react-native/src/NativeModule/office/debugpalette.json:1]: Makes sense. I definitely want your feedback on whether I am using the native module correctly. In particular I think I may be conflating several things. The idea of the base theme name and the palette that gets looked up from getPalette() as I think I'm using the base theme name as the default lookup value for the palette. But it seems like one of them should be things like 'Colorful' and the other should be things like 'TaskPane'.

Comment (FalseLobster) [reply packages/framework/theming-react-native/src/NativeModule/office/debugpalette.json:1]: Oh you're right I didn't even notice that at first.

File: packages/theming/win32-theme/src/createOfficeTheme.ts
Comment (FalseLobster) [inline packages/theming/win32-theme/src/createOfficeTheme.ts:23]: As you already noted initialHostThemeSetting is akin to Colorful/Black/Dark/Fresh whereas paletteName is basically a scheme within the broader office theme.

Comment (FalseLobster) [reply packages/theming/win32-theme/src/createOfficeTheme.ts:23]: paletteName could just be left undefined when passed into handlePaletteCall

File: packages/theming/theme-types/src/Theme.types.ts
Comment (FalseLobster) [inline packages/theming/theme-types/src/Theme.types.ts:25]: You don't have any quibbles about leaving this in type in a semi-official capacity?

File: packages/theming/win32-theme/src/createOfficeTheme.ts
Comment (FalseLobster) [reply packages/theming/win32-theme/src/createOfficeTheme.ts:23]: Since the themeRef ThemeTranslation closure should capture the paletteName, and the paletteName can be left undefined if the user didn't given, you can just replace that key value pair in the ref object with the initialHostThemeSetting / args.hostThemeSetting and forward that into createPartialTheme, where you can set
{
name: hostThemeSetting || '';
}

File: packages/experimental/theme/src/themeReference.ts
Comment (FalseLobster) [inline packages/experimental/theme/src/themeReference.ts:41]: Do we want to export some sort of baselineTheme ThemeReference?

Comment (FalseLobster) [reply packages/experimental/theme/src/themeReference.ts:41]: Neveeeeermind, I was just trying to avoid the boilerplate of 

`new ThemeReference(getBaselinePlatformTheme(), createOfficeTheme(), myThemeRecipe);`

but then I realized/remembered the createOfficeTheme was a ThemeReference not a ThemeRecipe and  is already layered on the default-theme.

I do wonder if we should take an optional baseThemeReference in createOfficeTheme in case they want to change that though,

Comment (FalseLobster) [reply packages/experimental/theme/src/themeReference.ts:41]: I'm not sure how common that scenario is in practice though.

**Review by FalseLobster — APPROVED** [review]
Overall this does seem friendlier to use, I was never a fan of the coupling I had to do between the ThemeRegistry and ThemeModuleHelper to get the old stuff working correctly.

File: packages/theming/theme-types/src/Theme.types.ts
Comment (JasonVMo) [reply packages/theming/theme-types/src/Theme.types.ts:25]: I've been debating this but I don't think so, though perhaps we rename it officePalette. I kind of envision host maybe eventually looking something like:

```ts
// host is basically a grouping for settings that come from native modules
host: {
  dark?: boolean;
  highContrast?: boolean;
  officePalette?: OfficePalette;
  win32?: Win32Settings;
  windows?: WindowsSettings;
  macos?: MacSettings;
  // or something like this
}
```

Then the types in ThemeReference are general so that someone can do the following:

```ts
type MyTheme = Theme & {
  host: {
    myAdditionalNativeSettings?: SomeType;
  }
}
```

Then basically fill those settings the same way the win32 theme does. You can imagine comments creating a theme, based off the Win32 theme, where they augment it with their native module but use our context and provider mechanism. Then we get rid of the dual context problem we've had in the past.

Comment (tkamireh) [reply packages/theming/theme-types/src/Theme.types.ts:25]: +1 to allowing for flexible and general usage, but one thing I'd be interested in is keeping all these names generic (`officePalette` seems a little constraining maybe) - there could be a world in the future where the native modules are written in a way where they're pluggable. Other brownfield scenarios where large applications serve as "hosts", aren't too far-fetched a notion imo.

Comment (JasonVMo) [reply packages/theming/theme-types/src/Theme.types.ts:25]: Yeah, I tend to agree though the question is where that coalesces. There are several levels of definitions and naming.

- **Base** - need a good name but these are things like `theme.colors.neutralPrimary` or `theme.colors.neutral3` or whatever we come up with
- **Semantic** - these are role based.  Right now they are things like `theme.colors.bodyBackground` but may change. Design is working on a new set of these now, and while they feel like the base layer may be volatile, these names might be stable(ish)
- **Native/Host** - at some point there is a different concept of naming that maps to what exists in native. That can either be exposed in the native module, or on the theme somewhere, or be hidden on the native side.

I think there are a couple of principles I've been thinking about:

- components should be authored against the semantic level. It is kind of the goldilocks level here and allows for a role based mapping of arbitrary color schemes.
- this means that customizations get authored against the semantic level, which makes sense
- this also implies native needs to map to semantic, rather than mapping to the base layer, which is great because this maps to the way these things tend to be built on the native side.
- this native -> semantic mapping can be lossy, meaning that some native values may have no mapping in the semantic layer. 
- This implies that there will likely be value in having those available somewhere in the theme, because then a component can be customized easily to depend on those colors on a given platform.

This is all subject to debate and discussion but this is just the way I've been starting to think about things.

File: packages/experimental/theme/src/themeReference.ts
Comment (JasonVMo) [reply packages/experimental/theme/src/themeReference.ts:41]: It's an interesting idea but I'm not sure how common it is in practice as well. It's easy enough to do, and we can do it later without changing the interface by factoring out a helper and exporting the more generic helper.  So I'm going to hold off for now.

File: packages/theming/win32-theme/src/createOfficeTheme.ts
Comment (JasonVMo) [reply packages/theming/win32-theme/src/createOfficeTheme.ts]: I updated the code around this to match the expected behavior in the latest iteration.

File: packages/framework/theming-react-native/src/NativeModule/ThemingModuleHelpers.ts
Comment (FalseLobster) [inline packages/framework/theming-react-native/src/NativeModule/ThemingModuleHelpers.ts:70]: You can't rename this without native theming module changes:

```
        dynamic params = dynamic::array("onPlatformDefaultsChanged", dynamic::object("hostThemeSetting", getThemeSetting()));
        instance->callJSFunction("RCTDeviceEventEmitter", "emit", std::move(params));
```
That creates a JSON object `{ hostThemeSetting: 'Foo' }`

Sorry, name is probably clear, I think the native code calls it the Theme Setting so I was probably borrowing that nomenclature

Comment (FalseLobster) [reply packages/framework/theming-react-native/src/NativeModule/ThemingModuleHelpers.ts:70]: Thankfully we can name it whatever we want in our pure JS interfaces, so this is really just an implementation detail

File: packages/theming/win32-theme/src/createOfficeTheme.ts
Comment (FalseLobster) [inline packages/theming/win32-theme/src/createOfficeTheme.ts:33]: Same here, the args object's shape is defined by the native side.

File: packages/theming/win32-theme/src/NativeModule/getThemingModule.native.ts
Comment (FalseLobster) [inline packages/theming/win32-theme/src/NativeModule/getThemingModule.native.ts:26]: So this does leave the NativeModule intact like the old version did, but you're still using a different fallbackGetPalette.

The old version did a require on the debugpalette.json you deleted.  That json was more fully specified than the impl in the fallbackThemingModule code, and it means that key-value pairs in the palette will disappear if we don't update the fallbackGetPalette to include the rest of the entries in the .json.

Now that I'm thinking about it, though, maybe we should get rid of the .json and export it from the NM as a constant.  I think I probably did it this way to avoid a second check in to the native code at the time.

File: packages/theming/theme-types/src/Theme.types.ts
Comment (tkamireh) [reply packages/theming/theme-types/src/Theme.types.ts:25]: I generally agree with the laid out principles. I would add, as a direct consequence of native -> semantic mapping be lossy + the probable usage of base when semantic is insufficient, that affordances for native -> base mappings are probably unavoidable. And I think the code already reflects this.

File: packages/framework/theming-react-native/src/NativeModule/office/debugpalette.json
Comment (JasonVMo) [reply packages/framework/theming-react-native/src/NativeModule/office/debugpalette.json:1]: Are we sure the debug palette is actually more complete? I believe the one defined in JS is fully specified since all the values in the palette are required.

File: packages/theming/win32-theme/src/NativeModule/getThemingModule.native.ts
Comment (JasonVMo) [reply packages/theming/win32-theme/src/NativeModule/getThemingModule.native.ts:26]: I think both the json and the code in the fallback have all the keys specified, there's an open question of which one is more correct, but they both should work.

File: packages/theming/win32-theme/src/createOfficeTheme.ts
Comment (JasonVMo) [reply packages/theming/win32-theme/src/createOfficeTheme.ts:33]: Good catch

File: packages/framework/theming-react-native/src/NativeModule/office/debugpalette.json
Comment (JasonVMo) [reply packages/framework/theming-react-native/src/NativeModule/office/debugpalette.json:1]: Actually looking at just the counts it looks like debugpalette does in fact have about 30 more entries. If that is complete, then we need to add those entries to the type in OfficePalette and then ensure they are specified.

Comment (JasonVMo) [reply packages/framework/theming-react-native/src/NativeModule/office/debugpalette.json:1]: I'll diff it later this afternoon

Comment (JasonVMo) [reply packages/framework/theming-react-native/src/NativeModule/office/debugpalette.json:1]: I added the missing values to the type and to the definition. There is one discrepancy, in the JSON file there are a series of values named things like StrokeOverRest and in the OfficePalette type they are named StrokeOverlayRest. I'm working off the assumption the types with Overlay were correct, not the Over ones.

File: packages/theming/win32-theme/src/NativeModule/getThemingModule.native.ts
Comment (JasonVMo) [reply packages/theming/win32-theme/src/NativeModule/getThemingModule.native.ts:26]: I did a diff of the files and added the types to the OfficePalette type, then patched the keys from the debugjson.json file. We probably do want to copy a full standard White or Colorful theme definition to ensure that when the fallbacks happen they are doing the right thing.

File: packages/theming/theme-types/src/Theme.types.ts
Comment (FalseLobster) [reply packages/theming/theme-types/src/Theme.types.ts:25]: Yea I think I'm _mostly_ on board with this strategy, although I think the one place I might object to the mapping of native -> semantic is with the brand/accent colors.

Comment (FalseLobster) [reply packages/theming/theme-types/src/Theme.types.ts:25]: Sent too soon --- I think you had concerns about this mapping being difficult to reason about because (currently) there are so many different accent colors in the ramp and we only use a subset--- I.e. if we prune out Accent 20-45, is it weird they we assign App2 to Accent 15 and App3 to Accent50?

I'm concerned that not mapping App -> Base layer here undercuts the flexibility of the Fluent theme and limits the ability to inject different brands.  Maybe this is something design should have a say in though.   Any thoughts about that?

File: packages/framework/theming-react-native/src/NativeModule/office/debugpalette.json
Comment (FalseLobster) [reply packages/framework/theming-react-native/src/NativeModule/office/debugpalette.json:1]: Actually the Over ones are correct--- the .JSON was taken straight off a devmain checkpoint the mock module was hand authored.

Other than that, the latest revision looks good to me.

File: packages/theming/theme-types/src/Theme.types.ts
Comment (JasonVMo) [reply packages/theming/theme-types/src/Theme.types.ts:25]: I definitely want design to iterate a bit more. From talking to Travis it sounds like their notion of base colors are more fluid than their notion of semantic ones. (They call them Sets) In general design might retarget the set value from Neutral.89 to Neutral.73 and we shouldn't rely on meaning for Neutral.73.

Because the set/semantic colors have explicit branded values that would be where we would map brand colors.

File: packages/framework/theming-react-native/src/NativeModule/office/debugpalette.json
Comment (JasonVMo) [reply packages/framework/theming-react-native/src/NativeModule/office/debugpalette.json:1]: Perfect, I updated the types, then changed the values and removed the commented out ones. I'm going to merge this PR once the CI loop passes. Thanks for all the feedback.

**Pull request merged**
